pkg/registry/generic/etcd/etcd.go

```diff
@@ -455,3 +444,18 @@ func (e *Etcd) WatchPredicate(ctx api.Context, m generic.Matcher, resourceVersio
 
 	return e.Helper.WatchList(e.KeyRootFunc(ctx), version, filterFunc)
 }
+
+// calculateTTL is a helper for retrieving the updated TTL for an object or returning an error
+// if the TTL cannot be calculated. The defaultTTL is changed to zero if less than zero.
+func (e *Etcd) calculateTTL(obj runtime.Object, defaultTTL int64, update bool) (ttl uint64, err error) {
+	// etcd may return a negative TTL for a node if the expiration has not occured due
+	// to server lag.
```

Comment:
It does not - I don't actually think that you can clear a 0 second TTL (the correct way to clear it from the call is to pass empty string, but I don't know what etcd server is doing yet, and the client ALWAYS sends the TTL).

----- Original Message -----
> > @@ -455,3 +444,18 @@ func (e *Etcd) WatchPredicate(ctx api.Context, m
> > generic.Matcher, resourceVersio
> >  
> >  	return e.Helper.WatchList(e.KeyRootFunc(ctx), version, filterFunc)
> >  }
> > +
> > +// calculateTTL is a helper for retrieving the updated TTL for an object
> > or returning an error
> > +// if the TTL cannot be calculated. The defaultTTL is changed to zero if
> > less than zero.
> > +func (e *Etcd) calculateTTL(obj runtime.Object, defaultTTL int64, update
> > bool) (ttl uint64, err error) {
> > +	// etcd may return a negative TTL for a node if the expiration has not
> > occured due
> > +	// to server lag.
> 
> Yeah, that SGTM. Unless it accepts a negative value, which I assume it
> doesn't if you're using a uint64.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/8724/files#r31066381